condor_history_to_prometheus.py

```diff
@@ -0,0 +1,192 @@
+            last_job[s.get('Name')] = {'ClusterId': None, 'EnteredCurrentStatus': None}
+
+    return last_job
+    
+
+def locate_schedds(collector):
+    try:
+        coll = htcondor.Collector(collector)
+        return coll.locateAll(htcondor.DaemonTypes.Schedd)
+    except htcondor.HTCondorIOError as e:
+        failed = e
+        logging.error(f'Condor error: {e}')
+
+def compose_ad_metrics(ad, metrics):
+    ''' Parse condor job classad and update metrics
+
+        Args:
+            ad (classad): an HTCondor job classad
+            metrics (JobMetrics): JobMetrics object 
+    '''
+    # ignore this ad if walltimehrs is negative or a dagman
+    if ad['walltimehrs'] < 0 or ad['Cmd'] == '/usr/bin/condor_dagman':
+        return
+
+    labels = {'owner': None,
+              'site': None,
+              'schedd': None,
+              'GPUDeviceName': None,
+              'usage': None,
+              'kind': None}
+
+    labels['owner'] = ad['Owner']
+    labels['site'] = ad['site']
+    labels['schedd'] = ad['GlobalJobId'][0:ad['GlobalJobId'].find('#')]
+    labels['GPUDeviceName'] = None
+    
+    if ad['ExitCode'] == 0 and ad['ExitBySignal'] is False and ad['JobStatus'] == 4:
+        labels['usage'] = 'goodput'
+    else:
+        labels['usage'] = 'badput'
+
+    resource_hrs = 0
+    resource_request = 0
+
+    if ad['Requestgpus'] > 0:
+        labels['kind'] = 'GPU'
+        labels['GPUDeviceName'] = ad['MachineAttrGPUs_DeviceName0']
+        resource_hrs = ad['gpuhrs']
+        resource_request = ad['Requestgpus']
+    else:
+        labels['kind'] = 'CPU'
+        resource_hrs = ad['cpuhrs']
+        resource_request = ad['RequestCpus']
+
+    metrics.condor_job_count.labels(**labels).inc()
+    metrics.condor_job_walltime_hours.labels(**labels).inc(ad['walltimehrs'])
+    metrics.condor_job_resource_hours.labels(**labels).inc(resource_hrs)
+    metrics.condor_job_resource_req.labels(**labels).observe(resource_request)
+    metrics.condor_job_mem_req.labels(**labels).observe(ad['RequestMemory']/1024)
+    metrics.condor_job_mem_used.labels(**labels).observe(ad['ResidentSetSize_RAW']/1048576)
+
+def query_collector(collector, metrics, last_job):
+    """Query schedds for job ads
+
+    Args:
+        collector (str): address for a collector to query
+        metrics (JobMetrics): JobMetrics instance
+        last_job (dict): dictionary for tracking last ClusterId by schedd
+    """
+    for schedd_ad in locate_schedds(collector):
+        name = schedd_ad.get('Name')
+
+        ads = read_from_schedd(schedd_ad, history=True, since=last_job[name]['ClusterId'])
+        if last_job[name]['EnteredCurrentStatus'] is not None:
+            logging.info(f'{name} - read ads since {last_job[name]["ClusterId"]}:{last_job[name]["EnteredCurrentStatus"]} at timestamp {datetime.strptime(last_job[name]["EnteredCurrentStatus"],utc_format)}')
+
+        for ad in generate_ads(ads):
+            if last_job[name]['ClusterId'] is None:
```

Comment:
Looking at how `last_job` is set, won't this only ever trigger once, on the first run through?  But we actually want to set it every time we call `query_collector()`, with the last job for each schedd.

Reply:
So, I set the `last_job` dict for the schedd if it's set to none, if it isn't then I do a comparison in the following line to see if that job ad has a more recent `EnteredCurrentStatus` than what is current stored in `last_job`. If it is, store that as the last job.

```
if datetime.strptime(ad['EnteredCurrentStatus'],utc_format) > datetime.strptime(last_job[name]['EnteredCurrentStatus'],utc_format):
```

I'm actually not entirely sure how necessary this is, `history()` normally returns jobs in the order of most recent to oldest in `EnteredCurrentStatus` but it more just to make sure because it doesn't seem to be guaranteed. so I err on the safe side.

Reply:
How ever, this did make me realize `last_job` should be a global for the script

Reply:
Ugh, I thought I was missing something.  I blame Friday afternoon.

Passing it to the function is fine, because it's a dict (so it gets passed by reference).